# Task Exercise 4.9 - Book 6. Creating an  Instacart sample dataframe

## In order to conduct further analysis and develop visualizations regarding the new orders_products_merged dataframe, a sample dataframe will be created.

# This script contains the following points:

## 1. Import new orders_products_merged dataframe

## 2. Create a sample set


In [1]:
# import libraries

import pandas as pd
import numpy as np
import os

## 1. Import new orders_products_merged' dataframe

In [2]:
path = r'C:\Users\howl6\OneDrive\Certificates\CareerFoundry\Coursework\Data_Immersion\Chapter 4\Instacart Basket Analysis'

In [3]:
# import orders_products_merged dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path,'02_Data', 'Prepared_Data', 'orders_products_merged.pkl' ))

In [4]:
ords_prods_merge.dtypes

product_id                    int64
product_name                 object
aisle_id                      int64
department_id                 int64
prices                      float64
order_id                      int64
user_id                      object
order_number                  int64
orders_day_of_week            int64
order_time_of_day             int64
days_since_prior_order      float64
first_time_order               bool
add_to_cart_order             int64
reordered                     int64
price_range_loc              object
busiest_day                  object
busiest_days                 object
busiest_period_of_day        object
max_order                     int64
loyalty_flag                 object
mean_prices                 float64
spending_flag                object
median_days_prior_order     float64
frequency_flag               object
first_name                   object
surname                      object
gender                       object
state                       

In [6]:
ords_prods_merge.shape

(32404859, 34)

In [7]:
ords_prods_merge.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  1.198023e+01  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.956554e+02  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  9.999900e+04  3.421083e+06   

       order_number  orders_day_of_week  order_time_of_day  \
count  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.714230e+01        2.738867e+00       1.342515e+01   
std    1.753532e+01        2.090077e+00       4.246380e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.400000e+01        5.000000e+00       1.600000e+01   
max    9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered     max_order  \
count            3.032876e+07       3.240486e+07  3.240486e+07  3.240486e+07   
mean             1.110408e+01       8.352547e+00  5.895873e-01  3.305217e+01   
std              8.779064e+00       7.127071e+00  4.919087e-01  2.515525e+01   
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%              5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01   
50%              8.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01   
75%              1.500000e+01       1.100000e+01  1.000000e+00  4.700000e+01   
max              3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01   

        mean_prices  median_days_prior_order           age  num_of_dependants  \
count  3.240486e+07             3.240485e+07  3.240486e+07       3.240486e+07   
mean   1.198023e+01             1.039776e+01  4.946527e+01       1.501896e+00   
std    8.324227e+01             7.131754e+00  1.848558e+01       1.118865e+00   
min    1.000000e+00             0.000000e+00  1.800000e+01       0.000000e+00   
25%    7.387298e+00             6.000000e+00  3.300000e+01       1.000000e+00   
50%    7.824786e+00             8.000000e+00  4.900000e+01       2.000000e+00   
75%    8.254023e+00             1.300000e+01  6.500000e+01       3.000000e+00   
max    2.500542e+04             3.000000e+01  8.100000e+01       3.000000e+00   

             income  
count  3.240486e+07  
mean   9.943773e+04  
std    4.305727e+04  
min    2.590300e+04  
25%    6.700400e+04  
50%    9.661800e+04  
75%    1.279120e+05  
max    5.939010e+05

### The shortage in rows regarding days_since_prior_order was attributed previously to new customers.

## 2. Create a sample set

### Due to the size of the merged dataframe (32404859 rows x 34 columns), a random sample of the data frame will be used for  analysis and visualizations.

In [8]:
# Creating a pattern of random numbers (in this case, seed “4”). Creating a random sample using a 70/30 split (from NumPy, rows assigned less than 0.7 in one sample, greater in the other)

np.random.seed(4)
dev = np.random.rand(len(ords_prods_merge)) <= 0.7

In [9]:
# Store 70% of the sample in the dataframe 'big'

big = ords_prods_merge[dev]

In [12]:
big.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  2.268376e+07  2.268376e+07   2.268376e+07  2.268376e+07  2.268376e+07   
mean   2.559844e+04  7.120029e+01   9.921023e+00  1.198749e+01  1.710901e+06   
std    1.408432e+04  3.821014e+01   6.281597e+00  4.970498e+02  9.872423e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354200e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.561000e+05   
50%    2.530000e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711197e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565522e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  9.999900e+04  3.421083e+06   

       order_number  orders_day_of_week  order_time_of_day  \
count  2.268376e+07        2.268376e+07       2.268376e+07   
mean   1.714296e+01        2.739048e+00       1.342535e+01   
std    1.753497e+01        2.090113e+00       4.247224e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.400000e+01        5.000000e+00       1.600000e+01   
max    9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered     max_order  \
count            2.123034e+07       2.268376e+07  2.268376e+07  2.268376e+07   
mean             1.110337e+01       8.352656e+00  5.896000e-01  3.305403e+01   
std              8.778893e+00       7.127245e+00  4.919063e-01  2.515524e+01   
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%              5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01   
50%              8.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01   
75%              1.500000e+01       1.100000e+01  1.000000e+00  4.700000e+01   
max              3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01   

        mean_prices  median_days_prior_order           age  num_of_dependants  \
count  2.268376e+07             2.268376e+07  2.268376e+07       2.268376e+07   
mean   1.197704e+01             1.039759e+01  4.946619e+01       1.501875e+00   
std    8.352866e+01             7.131419e+00  1.848517e+01       1.118916e+00   
min    1.000000e+00             0.000000e+00  1.800000e+01       0.000000e+00   
25%    7.387500e+00             6.000000e+00  3.300000e+01       1.000000e+00   
50%    7.824806e+00             8.000000e+00  4.900000e+01       2.000000e+00   
75%    8.254241e+00             1.300000e+01  6.500000e+01       3.000000e+00   
max    2.500542e+04             3.000000e+01  8.100000e+01       3.000000e+00   

             income  
count  2.268376e+07  
mean   9.943438e+04  
std    4.304513e+04  
min    2.590300e+04  
25%    6.700800e+04  
50%    9.660800e+04  
75%    1.279020e+05  
max    5.939010e+05

In [14]:
# export big sample dataframe

big.to_pickle(os.path.join(path,'02_Data','Prepared_Data', 'big_cop_sample.pkl'))

In [10]:
# Store 30% of the sample in the dataframe 'small'

small = ords_prods_merge[~dev]

In [15]:
small.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  9.721098e+06  9.721098e+06   9.721098e+06  9.721098e+06  9.721098e+06   
mean   2.559919e+04  7.118637e+01   9.916921e+00  1.196327e+01  1.710382e+06   
std    1.408325e+04  3.821429e+01   6.281223e+00  4.923865e+02  9.874306e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.555800e+05   
50%    2.530900e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.710702e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565431e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  9.999900e+04  3.421083e+06   

       order_number  orders_day_of_week  order_time_of_day  \
count  9.721098e+06        9.721098e+06       9.721098e+06   
mean   1.714078e+01        2.738444e+00       1.342467e+01   
std    1.753615e+01        2.089993e+00       4.244412e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.400000e+01        5.000000e+00       1.600000e+01   
max    9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered     max_order  \
count            9.098421e+06       9.721098e+06  9.721098e+06  9.721098e+06   
mean             1.110574e+01       8.352290e+00  5.895577e-01  3.304783e+01   
std              8.779462e+00       7.126665e+00  4.919141e-01  2.515526e+01   
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%              5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01   
50%              8.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01   
75%              1.500000e+01       1.100000e+01  1.000000e+00  4.700000e+01   
max              3.000000e+01       1.430000e+02  1.000000e+00  9.900000e+01   

        mean_prices  median_days_prior_order           age  num_of_dependants  \
count  9.721098e+06             9.721096e+06  9.721098e+06       9.721098e+06   
mean   1.198767e+01             1.039815e+01  4.946312e+01       1.501945e+00   
std    8.257014e+01             7.132534e+00  1.848655e+01       1.118746e+00   
min    1.000000e+00             0.000000e+00  1.800000e+01       0.000000e+00   
25%    7.386957e+00             6.000000e+00  3.300000e+01       1.000000e+00   
50%    7.824688e+00             8.000000e+00  4.900000e+01       2.000000e+00   
75%    8.253789e+00             1.300000e+01  6.500000e+01       3.000000e+00   
max    1.500694e+04             3.000000e+01  8.100000e+01       3.000000e+00   

             income  
count  9.721098e+06  
mean   9.944555e+04  
std    4.308561e+04  
min    2.590300e+04  
25%    6.699100e+04  
50%    9.663000e+04  
75%    1.279160e+05  
max    5.939010e+05

In [16]:
small.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'first_time_order',
       'add_to_cart_order', 'reordered', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'mean_prices', 'spending_flag', 'median_days_prior_order',
       'frequency_flag', 'first_name', 'surname', 'gender', 'state', 'age',
       'date_joined', 'num_of_dependants', 'family_status', 'income',
       '_merge'],
      dtype='object')

In [17]:
# export small sample dataframe

small.to_pickle(os.path.join(path,'02_Data','Prepared_Data', 'small_cop_sample.pkl'))